HeartbeatLogging/Sources/AnyEncoder.swift

```diff
@@ -0,0 +1,35 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Foundation
+
+/// A type that can encode an `Encodable` object into `Data`.
+protocol AnyEncoder {
+  func encode<T>(_ value: T) throws -> Data where T: Encodable
+}
```

Comment:
So the only reason I did this was to inject a decoder (or encoder for `AnyEncoder`) into a `HeartbeatStorage` instance could easily be faked. https://github.com/firebase/firebase-ios-sdk/blob/6f1c4eb1187a9056ac9cc07a4f98102cab1467c2/HeartbeatLogging/Sources/HeartbeatStorage.swift#L51-L54

Here's a test doing so: https://github.com/firebase/firebase-ios-sdk/blob/6f1c4eb1187a9056ac9cc07a4f98102cab1467c2/HeartbeatLogging/Tests/Unit/HeartbeatStorageTests.swift#L133-L166

We spoke briefly offline about it, this may be a lot of extra code just to improve testability for a code path that we should be able to verify most of the correctness at compile time:
https://github.com/firebase/firebase-ios-sdk/blob/6f1c4eb1187a9056ac9cc07a4f98102cab1467c2/HeartbeatLogging/Sources/HeartbeatStorage.swift#L139-L143

I'm ok with removing the `Any*` types and cleaning up the tests that tested this behavior.

WDYT?

Reply:
Thanks for sharing! For decoding, maybe we could avoid it by having it return `Data` that doesn't match the expected format? Even something like `"BAD_DATA".data(using: .utf8)`